db/range_del_aggregator.cc

```diff
@@ -138,159 +139,80 @@ class CollapsedRangeDelMap : public RangeDelMap {
            ucmp_->Compare(iter_->first, parsed.user_key) <= 0);
     assert(std::next(iter_) == rep_.end() ||
            ucmp_->Compare(parsed.user_key, std::next(iter_)->first) < 0);
-    return parsed.sequence < iter_->second.seq_;
+    return parsed.sequence < iter_->second;
   }
 
   bool IsRangeOverlapped(const Slice&, const Slice&) {
     // so far only implemented for non-collapsed mode since file ingestion (only
     // client) doesn't use collapsing
     assert(false);
+    return false;
   }
 
-  void AddTombstone(RangeTombstone tombstone) {
-    // In collapsed mode, we only fill the seq_ field in the RangeDelMap's
-    // values. The end_key is unneeded because we assume the tombstone extends
-    // until the next tombstone starts. For gaps between real tombstones and
-    // for the last real tombstone, we denote end keys by inserting fake
-    // tombstones with sequence number zero.
-    std::vector<RangeTombstone> new_range_dels{
-        tombstone, RangeTombstone(tombstone.end_key_, Slice(), 0)};
-    auto new_range_dels_iter = new_range_dels.begin();
-    // Position at the first overlapping existing tombstone; if none exists,
-    // insert until we find an existing one overlapping a new point
-    const Slice* rep_begin = nullptr;
-    if (!rep_.empty()) {
-      rep_begin = &rep_.begin()->first;
-    }
-    auto last_range_dels_iter = new_range_dels_iter;
-    while (new_range_dels_iter != new_range_dels.end() &&
-           (rep_begin == nullptr ||
-            ucmp_->Compare(new_range_dels_iter->start_key_, *rep_begin) < 0)) {
-      rep_.emplace(new_range_dels_iter->start_key_,
-                   RangeTombstone(Slice(), Slice(), new_range_dels_iter->seq_));
-      last_range_dels_iter = new_range_dels_iter;
-      ++new_range_dels_iter;
-    }
-    if (new_range_dels_iter == new_range_dels.end()) {
-      return;
+  void AddTombstone(RangeTombstone t) {
+    auto it = rep_.upper_bound(t.start_key_);
+    auto prev_seq = [&]() {
+      return it == rep_.begin() ? 0 : std::prev(it)->second;
+    };
+
+    // TODO: overall description of representation.
+    //
+    // Below, "a covers b" means that a has a greater sequence number than b.
+
+    // end_seq tracks the last sequence number we've covered with our new
+    // tombstone. It's the sequence number we'll install if we need to insert a
+    // new transition point for the new tombstone's end point.
+    SequenceNumber end_seq = 0;
+
+    if (t.seq_ > prev_seq()) {
```

Comment:
I feel you should add some diagrammatic comments here. For example:

```
// We're adding a new tombstone which is newer than an existing tombstone.
//
// 2: a-------m
// 1: a----------------z
//
// or
//
// 2: a----------------z
// 1: a-------m
//
// or
//
// 2:       m-----t
// 1: a----------------z
//
// or
//
// 2:         m--------z
// 1: a-----------t
//
// In all fource cases, we add an entry with the tombstone's start key. Note that in the latter two cases the existing
// tombstone's start key will cover the range to the new start key. We need to remember `end_seq` in 
// case the new tombstone does not completely cover the existing tombstones (the first and third cases).
```

Reply:
Done. Let me know if you find them confusing.